From d3d58ce5239b309773029f8f1a53547ceb56bf61 Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Sun, 19 Jan 2014 17:25:00 -0800
Subject: [PATCH] Check incompatible existing default symbol definition

After resolving a versioned reference, foo@VER1, to a default versioned
definition, foo@@VER1, from a shared object, we also merge it with
the existing regular default symbol definition, foo.  When foo is IFUNC
and foo@@VER1 aren't, we will merge 2 incompatible definitions.  This
patch avoids merging foo@@VER1 definition with foo definition if
one is IFUNC and the other isn't.
---
 bfd/ChangeLog                      |  8 +++++++
 bfd/elflink.c                      | 22 +++++++++++--------
 ld/testsuite/ChangeLog             | 14 ++++++++++++
 ld/testsuite/ld-ifunc/dummy.c      |  1 +
 ld/testsuite/ld-ifunc/ifunc.exp    | 45 ++++++++++++++++++++++++++++++++++++++
 ld/testsuite/ld-ifunc/pr16467.out  |  1 +
 ld/testsuite/ld-ifunc/pr16467a.c   |  5 +++++
 ld/testsuite/ld-ifunc/pr16467a.map |  4 ++++
 ld/testsuite/ld-ifunc/pr16467b.c   |  7 ++++++
 ld/testsuite/ld-ifunc/pr16467b.map |  4 ++++
 ld/testsuite/ld-ifunc/pr16467c.c   |  9 ++++++++
 11 files changed, 111 insertions(+), 9 deletions(-)
 create mode 100644 ld/testsuite/ld-ifunc/dummy.c
 create mode 100644 ld/testsuite/ld-ifunc/pr16467.out
 create mode 100644 ld/testsuite/ld-ifunc/pr16467a.c
 create mode 100644 ld/testsuite/ld-ifunc/pr16467a.map
 create mode 100644 ld/testsuite/ld-ifunc/pr16467b.c
 create mode 100644 ld/testsuite/ld-ifunc/pr16467b.map
 create mode 100644 ld/testsuite/ld-ifunc/pr16467c.c

bfd/

2014-01-20  H.J. Lu  <hongjiu.lu@intel.com>

	PR ld/16467
	* elflink.c (_bfd_elf_merge_symbol): When types of the existing
	regular default symbol definition and the versioned dynamic
	symbol definition mismatch, skip the default symbol definition
	if one of them is IFUNC.

ld/testsuite/

2014-01-20  H.J. Lu  <hongjiu.lu@intel.com>

	PR ld/16467
	* ld-ifunc/dummy.c: New file.
	* ld-ifunc/pr16467.out: Likewise.
	* ld-ifunc/pr16467a.c: Likewise.
	* ld-ifunc/pr16467a.map: Likewise.
	* ld-ifunc/pr16467b.c: Likewise.
	* ld-ifunc/pr16467b.map: Likewise.
	* ld-ifunc/pr16467c.c: Likewise.

	* ld-ifunc/ifunc.exp (run_cc_link_tests): New.
	(run_ld_link_exec_tests): Run pr16467.

diff --git a/bfd/elflink.c b/bfd/elflink.c
index 7dcafd6..3771af1 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -1088,19 +1088,23 @@ _bfd_elf_merge_symbol (bfd *abfd,
   /* When we try to create a default indirect symbol from the dynamic
      definition with the default version, we skip it if its type and
      the type of existing regular definition mismatch.  We only do it
-     if the existing regular definition won't be dynamic.  */
+     if the existing regular definition won't be dynamic unless one
+     of them is STT_GNU_IFUNC.  */
   if (pold_alignment == NULL
-      && !info->shared
-      && !info->export_dynamic
-      && !h->ref_dynamic
       && newdyn
       && newdef
       && !olddyn
-      && (olddef || h->root.type == bfd_link_hash_common)
-      && ELF_ST_TYPE (sym->st_info) != h->type
-      && ELF_ST_TYPE (sym->st_info) != STT_NOTYPE
-      && h->type != STT_NOTYPE
-      && !(newfunc && oldfunc))
+      && ((!info->shared
+	   && !info->export_dynamic
+	   && !h->ref_dynamic
+	   && (olddef || h->root.type == bfd_link_hash_common)
+	   && ELF_ST_TYPE (sym->st_info) != h->type
+	   && ELF_ST_TYPE (sym->st_info) != STT_NOTYPE
+	   && h->type != STT_NOTYPE
+	   && !(newfunc && oldfunc))
+	  || (olddef
+	      && ((h->type == STT_GNU_IFUNC)
+		  != (ELF_ST_TYPE (sym->st_info) == STT_GNU_IFUNC)))))
     {
       *skip = TRUE;
       return TRUE;
diff --git a/ld/testsuite/ld-ifunc/dummy.c b/ld/testsuite/ld-ifunc/dummy.c
new file mode 100644
index 0000000..5c03287
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/dummy.c
@@ -0,0 +1 @@
+/* An empty file.  */
diff --git a/ld/testsuite/ld-ifunc/ifunc.exp b/ld/testsuite/ld-ifunc/ifunc.exp
index fb106c6..d7ff445 100644
--- a/ld/testsuite/ld-ifunc/ifunc.exp
+++ b/ld/testsuite/ld-ifunc/ifunc.exp
@@ -349,6 +349,42 @@ if { $verbose < 1 } {
     remote_file host delete "tmpdir/static_nonifunc_prog"
 }
 
+run_cc_link_tests [list \
+    [list \
+	"Build libpr16467a.so" \
+	"-shared -Wl,--version-script=pr16467a.map" \
+	"-fPIC" \
+	{ pr16467a.c } \
+	{} \
+	"libpr16467a.so" \
+    ] \
+    [list \
+	"Build libpr16467b.a" \
+	"" \
+	"-fPIC" \
+	{ pr16467b.c } \
+	{} \
+	"libpr16467b.a" \
+    ] \
+    [list \
+	"Build libpr16467b.so" \
+	"-shared tmpdir/pr16467b.o tmpdir/libpr16467a.so \
+	 -Wl,--version-script=pr16467b.map" \
+	"-fPIC" \
+	{ dummy.c } \
+	{} \
+	"libpr16467b.so" \
+    ] \
+    [list \
+	"Build libpr16467c.a" \
+	"" \
+	"" \
+	{ pr16467c.c } \
+	{} \
+	"libpr16467c.a" \
+    ] \
+]
+
 run_ld_link_exec_tests [] [list \
     [list \
 	"Common symbol override ifunc test 1a" \
@@ -368,6 +404,15 @@ run_ld_link_exec_tests [] [list \
 	"ifunc-common-1.out" \
 	"-g" \
     ] \
+    [list \
+	"Run pr16467" \
+	"tmpdir/pr16467c.o tmpdir/libpr16467b.so tmpdir/libpr16467a.so" \
+	"" \
+	{ dummy.c } \
+	"pr16467" \
+	"pr16467.out" \
+	"" \
+    ] \
 ]
 
 set test_list [lsort [glob -nocomplain $srcdir/$subdir/*.d]]
diff --git a/ld/testsuite/ld-ifunc/pr16467.out b/ld/testsuite/ld-ifunc/pr16467.out
new file mode 100644
index 0000000..d86bac9
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/pr16467.out
@@ -0,0 +1 @@
+OK
diff --git a/ld/testsuite/ld-ifunc/pr16467a.c b/ld/testsuite/ld-ifunc/pr16467a.c
new file mode 100644
index 0000000..ae3f084
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/pr16467a.c
@@ -0,0 +1,5 @@
+const char * 
+sd_get_seats(void)
+{
+  return "OK";
+}
diff --git a/ld/testsuite/ld-ifunc/pr16467a.map b/ld/testsuite/ld-ifunc/pr16467a.map
new file mode 100644
index 0000000..d677f37
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/pr16467a.map
@@ -0,0 +1,4 @@
+LIBSYSTEMD_209 {
+global:
+        sd_get_seats;
+};
diff --git a/ld/testsuite/ld-ifunc/pr16467b.c b/ld/testsuite/ld-ifunc/pr16467b.c
new file mode 100644
index 0000000..264f6cf
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/pr16467b.c
@@ -0,0 +1,7 @@
+void new_sd_get_seats(void);
+__asm__(".symver new_sd_get_seats,sd_get_seats@LIBSYSTEMD_209");
+void (*resolve_sd_get_seats(void)) (void) __asm__ ("sd_get_seats");
+void (*resolve_sd_get_seats(void)) (void) {
+        return new_sd_get_seats;
+}
+__asm__(".type sd_get_seats, %gnu_indirect_function");
diff --git a/ld/testsuite/ld-ifunc/pr16467b.map b/ld/testsuite/ld-ifunc/pr16467b.map
new file mode 100644
index 0000000..1f263de
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/pr16467b.map
@@ -0,0 +1,4 @@
+LIBSYSTEMD_208 {
+global:
+        sd_get_seats;
+};
diff --git a/ld/testsuite/ld-ifunc/pr16467c.c b/ld/testsuite/ld-ifunc/pr16467c.c
new file mode 100644
index 0000000..e2a901c
--- /dev/null
+++ b/ld/testsuite/ld-ifunc/pr16467c.c
@@ -0,0 +1,9 @@
+#include <stdio.h>
+const char* sd_get_seats(void);
+
+int
+main (int argc, char **argv)
+{
+  printf("%s\n", sd_get_seats());
+  return 0;
+}
-- 
1.8.3.1

